File: iOS/Daily/Views/Health/SleepHypnogramView.swift
import SwiftUI
import DailyCore

/// Clinical 4-level hypnogram graphing Awake, REM, Light, and Deep stages along an hourly timeline.
public struct SleepHypnogramView: View {
    public let session: SleepSession
    @State private var selectedStage: SleepStageRecord? = nil
    
    public init(session: SleepSession) {
        self.session = session
    }
    
    private let rowHeight: CGFloat = 28
    private let laneSpacing: CGFloat = 8
    
    public var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            // Selected stage inspection tooltip
            if let selected = selectedStage {
                HStack(spacing: 8) {
                    Circle()
                        .fill(Color(hex: selected.stageType.hexColor))
                        .frame(width: 8, height: 8)
                    Text(selected.stageType.rawValue)
                        .font(.system(size: 13, weight: .bold))
                        .foregroundColor(.white)
                    Text("•")
                        .foregroundColor(ThemeColors.fgMutedDark)
                    Text("\(formatTime(selected.startTime)) - \(formatTime(selected.endTime))")
                        .font(.system(size: 12, weight: .medium))
                        .foregroundColor(.white.opacity(0.8))
                    Spacer()
                    Text("\(Int(round(selected.durationMinutes))) min")
                        .font(.system(size: 12, weight: .bold))
                        .foregroundColor(ThemeColors.accentCyan)
                }
                .padding(.horizontal, 12)
                .padding(.vertical, 6)
                .background(Capsule().fill(Color.white.opacity(0.08)))
                .transition(.opacity)
            }
            
            // Hypnogram Canvas
            GeometryReader { geo in
                let width = geo.size.width
                let totalSeconds = max(60, session.endTime.timeIntervalSince(session.startTime))
                
                ZStack(alignment: .topLeading) {
                    // 1. Horizontal lane guidelines & labels
                    VStack(alignment: .leading, spacing: laneSpacing) {
                        ForEach([SleepStageType.awake, .rem, .light, .deep], id: \.self) { stage in
                            HStack {
                                Text(stage.rawValue)
                                    .font(.system(size: 10, weight: .semibold))
                                    .foregroundColor(Color(hex: stage.hexColor).opacity(0.8))
                                    .frame(width: 44, alignment: .leading)
                                
                                Rectangle()
                                    .fill(Color.white.opacity(0.06))
                                    .frame(height: 1)
                            }
                            .frame(height: rowHeight)
                        }
                    }
                    
                    // 2. Vertical Hourly Grid Lines
                    let hours = generateHourlyTicks()
                    ForEach(hours, id: \.self) { hourDate in
                        let elapsed = hourDate.timeIntervalSince(session.startTime)
                        if elapsed >= 0 && elapsed <= totalSeconds {
                            let x = 48 + ((elapsed / totalSeconds) * (width - 52))
                            
                            Path { path in
                                path.move(to: CGPoint(x: x, y: 0))
                                path.addLine(to: CGPoint(x: x, y: 4 * (rowHeight + laneSpacing) - laneSpacing))
                            }
                            .stroke(Color.white.opacity(0.08), style: StrokeStyle(lineWidth: 1, dash: [3, 3]))
                        }
                    }
                    
                    // 3. Stage Blocks
                    ForEach(session.stages) { stage in
                        let elapsed = max(0, stage.startTime.timeIntervalSince(session.startTime))
                        let stageDur = max(30, stage.durationSeconds)
                        let left = 48 + ((elapsed / totalSeconds) * (width - 52))
                        let blockWidth = max(3, (stageDur / totalSeconds) * (width - 52))
                        let y = yPosition(for: stage.stageType)
                        let isSelected = selectedStage?.id == stage.id
                        
                        RoundedRectangle(cornerRadius: 4)
                            .fill(
                                LinearGradient(
                                    colors: [
                                        Color(hex: stage.stageType.hexColor).opacity(isSelected ? 1.0 : 0.85),
                                        Color(hex: stage.stageType.hexColor).opacity(isSelected ? 0.8 : 0.6)
                                    ],
                                    startPoint: .top,
                                    endPoint: .bottom
                                )
                            )
                            .overlay(
                                RoundedRectangle(cornerRadius: 4)
                                    .strokeBorder(isSelected ? Color.white : Color.white.opacity(0.2), lineWidth: isSelected ? 1.5 : 0.5)
                            )
                            .frame(width: blockWidth, height: rowHeight - 4)
                            .offset(x: left, y: y + 2)
                            .onTapGesture {
                                withAnimation(.spring(response: 0.3, dampingFraction: 0.7)) {
                                    if selectedStage?.id == stage.id {
                                        selectedStage = nil
                                    } else {
                                        selectedStage = stage
                                    }
                                }
                            }
                    }
                }
            }
            .frame(height: 4 * (rowHeight + laneSpacing) + 6)
            
            // X-Axis Time Ticks
            HStack {
                Text(session.bedtimeFormatted)
                    .font(.system(size: 11, weight: .bold, design: .rounded))
                    .foregroundColor(ThemeColors.fgMutedDark)
                Spacer()
                Text("← \(session.totalAsleepFormatted) asleep →")
                    .font(.system(size: 11, weight: .medium))
                    .foregroundColor(.white.opacity(0.6))
                Spacer()
                Text(session.wakeTimeFormatted)
                    .font(.system(size: 11, weight: .bold, design: .rounded))
                    .foregroundColor(ThemeColors.fgMutedDark)
            }
            .padding(.leading, 44)
        }
    }
    
    private func yPosition(for stage: SleepStageType) -> CGFloat {
        switch stage {
        case .awake: return 0
        case .rem: return rowHeight + laneSpacing
        case .light: return 2 * (rowHeight + laneSpacing)
        case .deep: return 3 * (rowHeight + laneSpacing)
        case .unknown: return 2 * (rowHeight + laneSpacing)
        }
    }
    
    private func generateHourlyTicks() -> [Date] {
        var ticks: [Date] = []
        let cal = Calendar.current
        var cur = cal.date(bySetting: .minute, value: 0, of: session.startTime) ?? session.startTime
        if cur < session.startTime {
            cur = cal.date(byAdding: .hour, value: 1, to: cur) ?? cur
        }
        while cur <= session.endTime {
            ticks.append(cur)
            cur = cal.date(byAdding: .hour, value: 1, to: cur) ?? cur.addingTimeInterval(3600)
        }
        return ticks
    }
    
    private func formatTime(_ date: Date) -> String {
        let f = DateFormatter()
        f.dateFormat = "HH:mm"
        return f.string(from: date)
    }
}

File: Android/app/src/main/java/com/intellidream/daily/presentation/health/SleepHypnogramView.kt
package com.intellidream.daily.presentation.health

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.intellidream.daily.designsystem.ThemeColors
import com.intellidream.daily.model.SleepSession
import com.intellidream.daily.model.SleepStageRecord
import com.intellidream.daily.model.SleepStageType
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import kotlin.math.max
import kotlin.math.roundToInt

@Composable
fun SleepHypnogramView(
    session: SleepSession,
    modifier: Modifier = Modifier
) {
    var selectedStage by remember { mutableStateOf<SleepStageRecord?>(null) }
    val timeFormatter = remember { SimpleDateFormat("HH:mm", Locale.getDefault()) }

    val rowHeightDp = 28.dp
    val laneSpacingDp = 8.dp

    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(12.dp)
    ) {
        // Selected stage inspection tooltip pill
        AnimatedVisibility(
            visible = selectedStage != null,
            enter = fadeIn(),
            exit = fadeOut()
        ) {
            selectedStage?.let { selected ->
                val stageColor = Color(android.graphics.Color.parseColor(selected.stageType.hexColor))
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clip(CircleShape)
                        .background(Color.White.copy(alpha = 0.08f))
                        .padding(horizontal = 12.dp, vertical = 6.dp),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    Box(
                        modifier = Modifier
                            .size(8.dp)
                            .clip(CircleShape)
                            .background(stageColor)
                    )
                    Text(
                        text = selected.stageType.rawValue,
                        fontSize = 13.sp,
                        fontWeight = FontWeight.Bold,
                        color = Color.White
                    )
                    Text(
                        text = "•",
                        fontSize = 12.sp,
                        color = ThemeColors.fgMutedDark
                    )
                    Text(
                        text = "${timeFormatter.format(Date(selected.startTime))} - ${timeFormatter.format(Date(selected.endTime))}",
                        fontSize = 12.sp,
                        fontWeight = FontWeight.Medium,
                        color = Color.White.copy(alpha = 0.85f)
                    )
                    Spacer(modifier = Modifier.weight(1f))
                    Text(
                        text = "${selected.durationMinutes.roundToInt()} min",
                        fontSize = 12.sp,
                        fontWeight = FontWeight.Bold,
                        color = ThemeColors.accentCyan
                    )
                }
            }
        }

        // Hypnogram Canvas
        val totalSeconds = max(60.0, (session.endTime - session.startTime) / 1000.0)
        val orderedStages = listOf(
            SleepStageType.AWAKE,
            SleepStageType.REM,
            SleepStageType.LIGHT,
            SleepStageType.DEEP
        )

        Canvas(
            modifier = Modifier
                .fillMaxWidth()
                .height(150.dp)
                .pointerInput(session.stages) {
                    detectTapGestures { offset ->
                        val w = size.width
                        val h = size.height
                        val leftOffsetPx = 48.dp.toPx()
                        val chartWidth = w - leftOffsetPx - 4.dp.toPx()
                        val rowH = 28.dp.toPx()
                        val laneSp = 8.dp.toPx()

                        if (offset.x >= leftOffsetPx) {
                            val elapsedSec = ((offset.x - leftOffsetPx) / chartWidth) * totalSeconds
                            val targetTime = session.startTime + (elapsedSec * 1000).toLong()

                            val hit = session.stages.firstOrNull { stage ->
                                targetTime in stage.startTime..stage.endTime
                            }

                            selectedStage = if (selectedStage?.id == hit?.id) null else hit
                        }
                    }
                }
        ) {
            val w = size.width
            val h = size.height
            val leftOffsetPx = 48.dp.toPx()
            val chartWidth = w - leftOffsetPx - 4.dp.toPx()
            val rowH = 28.dp.toPx()
            val laneSp = 8.dp.toPx()

            // 1. Draw horizontal guidelines for the 4 lanes
            for (idx in orderedStages.indices) {
                val laneY = idx * (rowH + laneSp) + (rowH / 2f)
                drawLine(
                    color = Color.White.copy(alpha = 0.08f),
                    start = Offset(leftOffsetPx, laneY),
                    end = Offset(w, laneY),
                    strokeWidth = 1.dp.toPx()
                )
            }

            // 2. Draw vertical hourly grid lines
            val cal = Calendar.getInstance().apply { timeInMillis = session.startTime }
            cal.set(Calendar.MINUTE, 0)
            cal.set(Calendar.SECOND, 0)
            if (cal.timeInMillis < session.startTime) {
                cal.add(Calendar.HOUR_OF_DAY, 1)
            }
            while (cal.timeInMillis <= session.endTime) {
                val elapsedSec = (cal.timeInMillis - session.startTime) / 1000.0
                val tickX = leftOffsetPx + ((elapsedSec / totalSeconds).toFloat() * chartWidth)
                drawLine(
                    color = Color.White.copy(alpha = 0.06f),
                    start = Offset(tickX, 0f),
                    end = Offset(tickX, 4 * (rowH + laneSp)),
                    strokeWidth = 1.dp.toPx()
                )
                cal.add(Calendar.HOUR_OF_DAY, 1)
            }

            // 3. Draw Stage Blocks
            for (stage in session.stages) {
                val elapsed = max(0.0, (stage.startTime - session.startTime) / 1000.0)
                val stageDur = max(30.0, stage.durationSeconds)
                val left = leftOffsetPx + ((elapsed / totalSeconds).toFloat() * chartWidth)
                val blockWidth = max(3f, (stageDur / totalSeconds).toFloat() * chartWidth)

                val laneIndex = when (stage.stageType) {
                    SleepStageType.AWAKE -> 0
                    SleepStageType.REM -> 1
                    SleepStageType.LIGHT, SleepStageType.UNKNOWN -> 2
                    SleepStageType.DEEP -> 3
                }
                val topY = laneIndex * (rowH + laneSp) + 2.dp.toPx()
                val blockHeight = rowH - 4.dp.toPx()

                val isSelected = selectedStage?.id == stage.id
                val stageColor = Color(android.graphics.Color.parseColor(stage.stageType.hexColor))

                // Block fill
                drawRoundRect(
                    color = stageColor.copy(alpha = if (isSelected) 1.0f else 0.85f),
                    topLeft = Offset(left, topY),
                    size = Size(blockWidth, blockHeight),
                    cornerRadius = CornerRadius(4.dp.toPx(), 4.dp.toPx())
                )

                // Block border
                drawRoundRect(
                    color = if (isSelected) Color.White else Color.White.copy(alpha = 0.2f),
                    topLeft = Offset(left, topY),
                    size = Size(blockWidth, blockHeight),
                    cornerRadius = CornerRadius(4.dp.toPx(), 4.dp.toPx()),
                    style = Stroke(width = if (isSelected) 1.5.dp.toPx() else 0.5.dp.toPx())
                )
            }
        }

        // X-Axis Time Ticks
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(start = 48.dp),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                text = session.bedtimeFormatted,
                fontSize = 11.sp,
                fontWeight = FontWeight.Bold,
                color = ThemeColors.fgMutedDark
            )

            Text(
                text = "← ${session.totalAsleepFormatted} asleep →",
                fontSize = 11.sp,
                fontWeight = FontWeight.Medium,
                color = Color.White.copy(alpha = 0.6f)
            )

            Text(
                text = session.wakeTimeFormatted,
                fontSize = 11.sp,
                fontWeight = FontWeight.Bold,
                color = ThemeColors.fgMutedDark
            )
        }
    }
}
